File: 04_10_Check_Subtree.cpp
class Solution {
public:
    
    bool isEqual(TreeNode* root, TreeNode* subRoot)
    {
        if ((root == nullptr) and (subRoot == nullptr))
            return true;

        else if (root == nullptr)
            return false;

        else if (subRoot == nullptr)
            return false;

        if (root->val != subRoot->val)
            return false;

        return (this->isEqual(root->left, subRoot->left)) and (this->isEqual(root->right, subRoot->right));
    }
    
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        
        if (root == nullptr)
            return false;
    
        if ((root->val == subRoot->val) and (this->isEqual(root, subRoot)))
            return true;
            
        return (this->isSubtree(root->right, subRoot)) or (this->isSubtree(root->left, subRoot)); 
    }
};

File: 04_10_Check_Subtree.java
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) {
            return true;
        }
        if (root == null || subRoot == null) {
            return false;
        }
        if (isSubtreeHelper(root, subRoot)) {
            return true;
        } else {
            return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
        }
    }
    
    boolean isSubtreeHelper(TreeNode current, TreeNode currentSubRoot) {
        if (current == null && currentSubRoot == null) {
            return true;
        }
        if (current == null || currentSubRoot == null) {
            return false;
        }
        
        if (current.val == currentSubRoot.val) {
            return isSubtreeHelper(current.left, currentSubRoot.left)
                && isSubtreeHelper(current.right, currentSubRoot.right);
        } else {
            return false;
        }
    }
}
